golem-worker-executor-base/src/grpc.rs

```diff
@@ -471,6 +472,56 @@ impl<Ctx: WorkerCtx, Svcs: HasAll<Ctx> + UsesAllDeps<Ctx = Ctx> + Send + Sync +
         }
     }
 
+    async fn cancel_invocation_internal(
+        &self,
+        request: CancelInvocationRequest,
+    ) -> Result<(), GolemError> {
+        let owned_worker_id =
+            extract_owned_worker_id(&request, |r| &r.worker_id, |r| &r.account_id)?;
+        self.ensure_worker_belongs_to_this_executor(&owned_worker_id)?;
+
+        let idempotency_key = request
+            .idempotency_key
+            .ok_or(GolemError::invalid_request("idempotency_key not found"))?
+            .into();
+
+        let metadata = Worker::<Ctx>::get_latest_metadata(&self.services, &owned_worker_id).await?;
+
+        match metadata {
+            Some(metadata) => {
+                if metadata
+                    .last_known_status
+                    .pending_invocations
+                    .iter()
+                    .any(|invocation| {
+                        invocation.invocation.idempotency_key() == Some(&idempotency_key)
+                    })
+                {
+                    let worker = Worker::get_or_create_suspended(
+                        self,
+                        &owned_worker_id,
+                        None,
+                        None,
+                        None,
+                        None,
+                    )
+                    .await?;
+                    worker.cancel_invocation(idempotency_key).await?;
+                    Ok(())
+                } else if metadata
+                    .last_known_status
+                    .invocation_results
+                    .contains_key(&idempotency_key)
+                {
+                    Err(GolemError::invalid_request("Invocation already completed"))
```

Comment:
is cancelling already completed an error? shouldn't it e.g. returned as bool?

Reply:
Yeah probably that's a better API as the user has no control in whether it has been already completed or not.
Should we change it?

Reply:
I think we should change it now, as this can be quite frustrating when comes up and somebody wants to apply same logic based on it.

Reply:
done